### DISCUSSION TITLE
Exact types of useQuery first argument for wrapper hook

### DISCUSSION THREAD

**Discussion by Sheraff** [trpc/trpc#2412] [Q&A] [answered]
I'm trying to make a wrapper around *every* call to `trpc.useQuery`. It should look something like this:

```js
function wrapper(pathAndInput) {
  const result = trpc.useQuery(pathAndInput)
  /* some code that also uses `pathAndInput` */
  return result
}
```

However I can't make work the fact that the `pathAndInput` can have a variadic second element. I.e. in `trpc.useQuery(["key", params], options)` the `params` part is optional.

The type I have so far is this:
```ts
function useIndexedTRcpQuery<TRouteKey extends TQuery>(
	pathAndInput: inferQueryInput<TRouteKey> extends (undefined | void | null)
		? [TRouteKey]
		: [TRouteKey, inferQueryInput<TRouteKey>]
): UseQueryResult<inferQueryOutput<TRouteKey>>
```

However I get the following typescript error:
> ```
> trpc.useQuery(pathAndInput)
>               ^^^^^^^^^^^^
> Source has 1 element(s) but target requires 2
> ```

I have also tried the following:
```ts
function useIndexedTRcpQuery<TRouteKey extends TQuery>(
	pathAndInput: [
		key: TRouteKey,
		args?: inferQueryInput<TRouteKey>,
	],
): UseQueryResult<inferQueryOutput<TRouteKey>>
```

Which gives the following error :
> ```
> trpc.useQuery(pathAndInput)
>               ^^^^^^^^^^^^
> Source provides no match for variadic element at position 1 in target
> ```

---

For reference, here are what the "utils types" are in the above code (pretty standard trpc stuff):
```ts
import type { UseQueryResult } from "react-query";
import type { inferProcedureOutput, inferProcedureInput } from "@trpc/server";

type TQuery = keyof AppRouter["_def"]["queries"];

type inferQueryInput<
  TRouteKey extends TQuery,
> = inferProcedureInput<AppRouter["_def"]["queries"][TRouteKey]>;

type inferQueryOutput<
  TRouteKey extends TQuery,
> = inferProcedureOutput<AppRouter["_def"]["queries"][TRouteKey]>;
```

**Comment by gabriel-frattini** [comment, marked as answer]
I think you want to import `inferHandlerInput`  from ` @trpc/server` in order to infer the input at runtime

This should give you a typed query with a optional `params` if it's needed for the query
```typescript
import type { inferHandlerInput } from '@trpc/server'

type TQuery = keyof AppRouter["_def"]["queries"];

type TPathAndArgs<TRouteKey extends TQuery> = [
    path: TRouteKey & string,
    ...args: inferHandlerInput<AppRouter['_def']['queries'][TRouteKey]>
  ]

  function useIndexedTRcpQuery<TRouteKey extends TQuery>(
    pathAndInput: TPathAndArgs<TRouteKey>
  ): : UseQueryResult<inferQueryOutput<TRouteKey>>

```

**Reply by Sheraff** [reply]
Thanks a ton!

I had tried very similar things but with `inferProcedureInput` instead of `inferHandlerInput`. The terms "procedure" and "handler" aren't very clear to me...
